packages/concerto-core/lib/model/customresource.js

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const { TypedStack } = require('@accordproject/concerto-util');
+const Resource = require('./resource');
+
+/**
+ * CustomResource is a Resource that is designed to be extended by
+ * generated client classes.
+ * @abstract
+ */
+class CustomResource extends Resource {
+    /**
+     * Constructor.
+     * @param {unknown} opaque Opaque data
+     */
+    constructor(opaque) {
+        // These properties are internals and should not be exposed to the caller.
+        const { modelManager, classDeclaration, ns, type, id, timestamp, resourceValidator } = opaque;
+        super(modelManager, classDeclaration, ns, type, id, timestamp);
+        this.$validator = resourceValidator;
+    }
+
+    /**
+     * Sets a property. If validation is enabled, then the property value is first validated
+     * to ensure that it does not violate the model.
+     * @param {string} propName - the name of the field
+     * @param {string} value - the value of the property
+     * @throws {Error} if the value is not compatible with the model definition for the field
+     */
+    setPropertyValue(propName, value) {
```

Comment:
This feels a little like we're doing optional inheritance between `Resource` and `ValidatedResource`. Is this the best pattern, here?

Reply:
@mttrbrts I think if I had the opportunity to make breaking changes I would have merged `Resource` and `ValidatedResource` (so that `Resource` simply has optional validation), changed the constructor signature for `Resource` (so that it accepts an "opaque" object) and just used `Resource` instead. 

Maybe a v3 item to revisit? Any other suggestions?

Reply:
I'm torn on this one... I think regretting those runtime Resource domain objects, and forcing people to use Factory and Serialiser to manipulate them. I think it is steepens the learning curve.

I think the [Concerto](https://github.com/accordproject/concerto/blob/master/packages/concerto-core/lib/concerto.js) functional class shows that we can accomplish almost everything with a functional API that sits on-top of JSON data, without paying the cost of JSON serialisation to/from domain objects.

Irrespective of this PR, what do you think would be the preferred way forward for Concerto v2?